File: graphs/a_star.js
/**
 * A* Search Algorithm
 */

class Node {
    constructor(x, y, parent = null) {
        this.x = x;
        this.y = y;
        this.parent = parent;
        this.g = 0;
        this.h = 0;
        this.f = 0;
    }
}

function aStar(grid, start, end) {
    let openList = [];
    let closedList = [];

    openList.push(start);

    while (openList.length > 0) {
        let lowInd = 0;
        for (let i = 0; i < openList.length; i++) {
            if (openList[i].f < openList[lowInd].f) {
                lowInd = i;
            }
        }

        let currentNode = openList[lowInd];

        if (currentNode.x === end.x && currentNode.y === end.y) {
            let curr = currentNode;
            let ret = [];
            while (curr.parent) {
                ret.push([curr.x, curr.y]);
                curr = curr.parent;
            }
            ret.push([curr.x, curr.y]);
            return ret.reverse();
        }

        openList.splice(lowInd, 1);
        closedList.push(currentNode);

        let neighbors = [];
        let x = currentNode.x;
        let y = currentNode.y;

        if (grid[x - 1] && grid[x - 1][y]) neighbors.push(new Node(x - 1, y, currentNode));
        if (grid[x + 1] && grid[x + 1][y]) neighbors.push(new Node(x + 1, y, currentNode));
        if (grid[x][y - 1] && grid[x][y - 1]) neighbors.push(new Node(x, y - 1, currentNode));
        if (grid[x][y + 1] && grid[x][y + 1]) neighbors.push(new Node(x, y + 1, currentNode));

        for (let i = 0; i < neighbors.length; i++) {
            let neighbor = neighbors[i];

            if (closedList.find(n => n.x === neighbor.x && n.y === neighbor.y) || grid[neighbor.x][neighbor.y] === 0) {
                continue;
            }

            let gScore = currentNode.g + 1;
            let gScoreIsBest = false;

            if (!openList.find(n => n.x === neighbor.x && n.y === neighbor.y)) {
                gScoreIsBest = true;
                neighbor.h = Math.abs(neighbor.x - end.x) + Math.abs(neighbor.y - end.y);
                openList.push(neighbor);
            } else if (gScore < neighbor.g) {
                gScoreIsBest = true;
            }

            if (gScoreIsBest) {
                neighbor.parent = currentNode;
                neighbor.g = gScore;
                neighbor.f = neighbor.g + neighbor.h;
            }
        }
    }
    return [];
}

// Driver Code
// 1 = Walkable, 0 = Wall
let grid = [
    [1, 0, 1, 1, 1],
    [1, 1, 1, 0, 1],
    [1, 1, 1, 0, 1],
    [0, 0, 1, 0, 1],
    [1, 1, 1, 0, 1]
];

let startNode = new Node(0, 0);
let endNode = new Node(4, 4);

let path = aStar(grid, startNode, endNode);
console.log("Path:", path);

File: graphs/a_star.go
package main

import (
	"container/heap"
	"fmt"
	"math"
)

type Point struct {
	x, y int
}

type Node struct {
	pt     Point
	g, h, f int
	parent *Node
	index  int
}

type PriorityQueue []*Node

func (pq PriorityQueue) Len() int { return len(pq) }

func (pq PriorityQueue) Less(i, j int) bool {
	return pq[i].f < pq[j].f
}

func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}

func (pq *PriorityQueue) Push(x interface{}) {
	n := len(*pq)
	item := x.(*Node)
	item.index = n
	*pq = append(*pq, item)
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	item.index = -1
	*pq = old[0 : n-1]
	return item
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func heuristic(a, b Point) int {
	return abs(a.x-b.x) + abs(a.y-b.y)
}

func AStar(grid [][]int, start, end Point) []*Point {
	rows := len(grid)
	cols := len(grid[0])

	openList := make(PriorityQueue, 0)
	heap.Init(&openList)

	startNode := &Node{pt: start, g: 0, h: 0, f: 0}
	heap.Push(&openList, startNode)

	closedList := make(map[Point]bool)
	
	// Keep track of visited nodes to avoid re-adding
	nodeMap := make(map[Point]*Node)
	nodeMap[start] = startNode

	for openList.Len() > 0 {
		current := heap.Pop(&openList).(*Node)

		if current.pt == end {
			var path []*Point
			for curr := current; curr != nil; curr = curr.parent {
				path = append(path, &curr.pt)
			}
			// Reverse
			for i, j := 0, len(path)-1; i < j; i, j = i+1, j-1 {
				path[i], path[j] = path[j], path[i]
			}
			return path
		}

		closedList[current.pt] = true

		moves := []Point{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}
		for _, move := range moves {
			neighborPt := Point{current.pt.x + move.x, current.pt.y + move.y}

			if neighborPt.x < 0 || neighborPt.x >= rows || neighborPt.y < 0 || neighborPt.y >= cols {
				continue
			}

			if grid[neighborPt.x][neighborPt.y] == 0 || closedList[neighborPt] {
				continue
			}

			gScore := current.g + 1
			hScore := heuristic(neighborPt, end)
			fScore := gScore + hScore

			if existingNode, exists := nodeMap[neighborPt]; exists {
				if gScore < existingNode.g {
					existingNode.g = gScore
					existingNode.f = fScore
					existingNode.parent = current
					heap.Fix(&openList, existingNode.index)
				}
			} else {
				neighborNode := &Node{
					pt:     neighborPt,
					g:      gScore,
					h:      hScore,
					f:      fScore,
					parent: current,
				}
				nodeMap[neighborPt] = neighborNode
				heap.Push(&openList, neighborNode)
			}
		}
	}
	return nil
}

func main() {
	// 1-Walkable, 0-Blocked
	grid := [][]int{
		{1, 0, 1, 1, 1},
		{1, 1, 1, 0, 1},
		{1, 1, 1, 0, 1},
		{0, 0, 1, 0, 1},
		{1, 1, 1, 0, 1},
	}

	path := AStar(grid, Point{0, 0}, Point{4, 4})
	if path != nil {
		fmt.Println("Path found:")
		for _, pt := range path {
			fmt.Printf("(%d, %d) ", pt.x, pt.y)
		}
		fmt.Println()
	} else {
		fmt.Println("No path found")
	}
}